### DISCUSSION TITLE
MIDI as part of a Composite Device

### DISCUSSION THREAD

**Discussion by PaulFortescue** [hathach/tinyusb#2162] [Q&A] [answered]
### Operating System

Windows 11

### Board

Raspberry Pi Pico

### Firmware

examples/device/msc is the base of this.

I can incorporate msc/cdc/hid into a composite device using the examples provided and it works perfectly.

I can use MIDI as a standalone device using the example provided and it works perfectly.

I cannot, using any permutation or combination of settings, make MIDI work concurrently with any other device or devices. Windows rejects the device with a (not very useful) code 10 meaning "something is wrong".

However, I deduce from the internet that it IS possible to incorporate MIDI as part of a composite device.

### What happened ?

I am only guessing that there is a bug somewhere that prevents me using TinyUSB to do this?

I've spent 3 days solid on it now so I'm hoping for an answer, even if it's "no"!

PS many, many thanks for the TinyUSB library!

### How to reproduce ?

Working with the "//" lines in, not with them out:

enum
{
	ITF_NUM_CDC = 0,
	ITF_NUM_CDC_DATA,
	ITF_NUM_MSC,
//	ITF_NUM_MIDI,
//	ITF_NUM_MIDI_STREAMING,
	ITF_NUM_TOTAL
};

#define CONFIG_TOTAL_LEN    (TUD_CONFIG_DESC_LEN + TUD_CDC_DESC_LEN + TUD_MSC_DESC_LEN) 
// + TUD_MIDI_DESC_LEN 

uint8_t const desc_fs_configuration[] =
{
	// Config number, interface count, string index, total length, attribute, power in mA
	TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),

	// Interface number, string index, EP notification address and size, EP data address (out, in) and size.
	TUD_CDC_DESCRIPTOR(ITF_NUM_CDC, 4, EPNUM_CDC_NOTIF, 8, EPNUM_CDC_OUT, EPNUM_CDC_IN, 64),

	// Interface number, string index, EP Out & EP In address, EP size
	TUD_MSC_DESCRIPTOR(ITF_NUM_MSC, 5, EPNUM_MSC_OUT, EPNUM_MSC_IN, 64),

//	TUD_MIDI_DESCRIPTOR(ITF_NUM_MIDI, 0, EPNUM_MIDI, 0x80 | EPNUM_MIDI, 64),

};

### Debug Log as txt file (LOG/CFG_TUSB_DEBUG=2)

//

### I have checked existing issues, dicussion and documentation

- [X] I confirm I have checked existing issues, dicussion and documentation.

**Comment by hathach** [comment]
midi work just fine with other class, try out arduino example here 
- https://github.com/adafruit/Adafruit_TinyUSB_Arduino/tree/master/examples/MIDI/midi_test

Arduino always keep the cdc therefore this is cdc + midi. You probably get the descriptor incorrectly. You can troubleshoot this by enabling debug log and/or dumping configuration and use online parse tool to spot any error.

Btw your how-to-reproduce is not reproducible, it is best to post a link to your example repo.

**Comment by PaulFortescue** [comment]
Hello Mr Thach and thank you very much for your quick reply!

I'm going to attempt to put my code on here, but I've never used GitHub except for looking at code, so bear with me.

I've solved my problem by having two different code sets which I will use one at a time, as I don't actually really need composite (I just need the same port to be either MIDI or CDC depending on a switch).

**Comment by PaulFortescue** [comment]
I've posted all the code at https://github.com/PaulFortescue/Miscellaneous

(I think)

If anyone wants to put me right I'm listening. I probably don't know what I'm doing, so please don't shout at me.

All I know is that the code WILL work if I take MIDI out, or WILL work if I ONLY leave MIDI in. See original post for details.

**Comment by hathach** [comment]
converted to discussion, since this is clearly not an bug from the stack.

**Reply by PaulFortescue** [reply]
👍thank you!

**Comment by samspencer5991** [comment]
I'm currently searching for a solution on this myself.
I'm porting my USB drivers from the ST32Cube Middlewares drivers to TinyUSB.
I've tested MIDI by itself, no issues. CDC, same thing. But combining the two leads to issues.

Fortunately I've code my previous codebase with the ST Middlewares to compare against, and checking both the descriptors in code, and also what is sent over wireshark, the issue isn't with my descriptors, they are identical to those in my ST driver application.

While debugging, an assert pauses indicating a driver issue:
`// Failed if there is no supported drivers
TU_ASSERT(drv_id < TOTAL_DRIVER_COUNT);`

**Reply by PaulFortescue** [reply]
I found exactly the same. I don't even know if it's possible to combine MIDI with another device as a composite, although a search on the Internet told me someone had found a way to do it (but hadn't posted the solution). Luckily (for me) I don't actually need to, all I need is to do one or the other.

I strongly suspect there is an issue with TinyUSB and I look forward to a solution. Good luck!

**Reply by samspencer5991** [reply]
Fortunately I know it's completely possible as I many many devices currently running that configuration using the ST stack.

**Reply by PaulFortescue** [reply]
I've not used the ST stack yet, although I know I will need to very soon on a U575. But not MIDI, just CDC I think. I hope you find your answer soon! Perhaps Mr Ha will agree with you that it's a problem with TinyUSB (if indeed it is)? Give me a heads-up if you find it - I'll do the same if I find it first :)

**Reply by samspencer5991** [reply]
I'll wait to see if @hathach has any thoughts. Let me know what you need from me. As I mentioned, fairly sure the descriptors are good, it only deviates from the example MIDI by removing the 2x 9 byte AC descriptors.

**Comment by sidprice** [comment]
I am starting to implement CDC and Midi so I am hoping to see some information here. At least someone confirming that CDC/Midi is a valid composite device with TinyUSB.

**Reply by samspencer5991** [reply]
I mean, I've only done it with the ST stack, and it seems to be a Tiny USB issue loading the embedded drivers in run time. No luck with the TinyUSB stack yet

**Comment by sidprice** [comment]
I really don't want to use the ST stack, hence investigating TinyUSB.

**Reply by samspencer5991** [reply]
Lol yes I know, I was just letting you know my previous experiences

**Comment by rppicomidi** [comment, marked as answer]
@PaulFortescue I haven't tried the [code you posted](https://github.com/PaulFortescue/Miscellaneous), but I did take a quick look and found what could be the issue. In the file `usb_descriptors.c`, find this section:
```
#define EPNUM_CDC_NOTIF   0x81
#define EPNUM_CDC_OUT     0x02
#define EPNUM_CDC_IN      0x82

#define EPNUM_MSC_OUT     0x03
#define EPNUM_MSC_IN      0x83

#define EPNUM_MIDI   0x01
```
It looks like the CDC notification endpoint conflicts with the MIDI endpoint. Have you tried changing
```
#define EPNUM_MIDI   0x01
```
to the next available unused endpoint number, which is 4?
```
#define EPNUM_MIDI   0x04
```
?

**Reply by PaulFortescue** [reply]
Hello rpipicomid

I've just tried that and the CDC is indeed working, however the MIDI seems to be no more. I'm pursuing it though, I forgot that the top bit signifies input and mistakenly thought that 0x01 was different from 0x81!

Thanks very much!

**Reply by PaulFortescue** [reply]
Hello again rpipicomid

You are indeed a king amonst men and an all-round top guy!

It's all working as it should be. I hope this helps Sam Spencer too, and anyone else who finds it!

Thanks again😊

**Comment by sidprice** [comment]
@PaulFortescue Did you update the code you posted earlier; I would like to see it an compare to mine :)

**Reply by PaulFortescue** [reply]
Hi Sid

The only bits that matter are:

```
tusb_desc_device_t const desc_device =
{
	.bLength = sizeof(tusb_desc_device_t),
	.bDescriptorType = TUSB_DESC_DEVICE,
	.bcdUSB = 0x0200,
	.bDeviceClass = TUSB_CLASS_MISC,
	.bDeviceSubClass = MISC_SUBCLASS_COMMON,
	.bDeviceProtocol = MISC_PROTOCOL_IAD,
	.bMaxPacketSize0 = CFG_TUD_ENDPOINT0_SIZE,

	.idVendor = 0xCafe,
	.idProduct = USB_PID,
	.bcdDevice = 0x0100,

	.iManufacturer = 0x01,
	.iProduct = 0x02,
	.iSerialNumber = 0x03,

	.bNumConfigurations = 0x01
};

enum
{
	ITF_NUM_CDC            = 0,
	ITF_NUM_CDC_DATA,
	ITF_NUM_MIDI,
	ITF_NUM_MIDI_STREAMING,
	ITF_NUM_TOTAL
};

#define CONFIG_TOTAL_LEN_MIDI  (TUD_CONFIG_DESC_LEN + TUD_MIDI_DESC_LEN + TUD_CDC_DESC_LEN)
//#define CONFIG_TOTAL_LEN_CDC  (TUD_CONFIG_DESC_LEN + TUD_CDC_DESC_LEN)


#define EPNUM_CDC_NOTIF 0x81
#define EPNUM_CDC_OUT   0x02
#define EPNUM_CDC_IN    0x82

#define EPNUM_MIDI   0x04

uint8_t const desc_fs_configuration_midi[] =
{
	// Config number, interface count, string index, total length, attribute, power in mA
	TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN_MIDI, TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 100),

	// Interface number, string index, EP Out & EP In address, EP size
	TUD_CDC_DESCRIPTOR(ITF_NUM_CDC, 4, EPNUM_CDC_NOTIF, 8, EPNUM_CDC_OUT, EPNUM_CDC_IN, 64),

	TUD_MIDI_DESCRIPTOR(ITF_NUM_MIDI, 0, EPNUM_MIDI, 0x80 | EPNUM_MIDI, 64),
		
};



```

Let me know if you need more?

**Comment by sidprice** [comment]
Thank you, I will give that a try when I switch back to that project.

**Reply by PaulFortescue** [reply]
I'm the same, I only dip into that one while I'm waiting for something on all of the others 😊

**Reply by samspencer5991** [reply]
I was actually able to solve me problem too, although it wasn't related to any endpoint issues. Strangely it was caused by using the STM32 VID. As soon as I changed the VID to anything else, no issues. The CDC worked regardless, but MIDI was not enumerating. Running on a STM32G4

**Comment by rppicomidi** [comment]
@PaulFortescue Glad I could help. Here is how I found the issue: EP 0x01, an OUT endpoint is indeed different from EP 0x81, an IN
endpoint. However, the MIDI device defined in the macro has both OUT and IN endpoints. In the original code 
with `#define EPNUM_MIDI   0x01`, the 4th argument in the line
```
TUD_MIDI_DESCRIPTOR(ITF_NUM_MIDI, 0, EPNUM_MIDI, 0x80 | EPNUM_MIDI, 64)
```
specifies IN EP 0x81 in the expression because if `EPNUM_MIDI == 1`, then
```
0x80 | EPNUM_MIDI == 0x80 | 0x01 == 0x81
```
.
